File: leetcode-swift/lc_233.swift
class Solution {
    func countDigitOne(_ n: Int) -> Int {
        var result = 0
        var position = 1

        while position <= n {
            var cycle = n / (position * 10)

            let currentDigit = (n / position) % 10
            if currentDigit > 1 {
                cycle += 1
            }

            result += cycle * position
            if currentDigit == 1 {
                result += n % position + 1
            }

            position *= 10
        }

        return result
    }
}

File: leetcode-python/lc_233.py
class Solution:
    def countDigitOne(self, n: int) -> int:
        if n <= 0:
            return 0
        
        m = n
        sum = 0
        e = 1
        while n > 0:
            r = n % 10
            n //= 10
            if r == 0:
                sum += n * e
            elif r > 1:
                sum += (n + 1) * e
            else: # r == 1
                sum += m - n * 9 * e - e + 1
                
            e *= 10
            
        return sum
